File: controllers/overflowController.go
package controllers

import (
	"net/http"
	"github.com/gin-gonic/gin"
	"Backend/models"
	"Backend/config"
	"context"
	"go.mongodb.org/mongo-driver/bson"
)

// Get all overflows
func GetOverflows(c *gin.Context) {
	var overflows []models.Overflow
	cursor, err := config.Client.Database("your_db").Collection("overflows").Find(context.TODO(), bson.M{})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Error fetching overflows"})
		return
	}
	defer cursor.Close(context.TODO())
	if err = cursor.All(context.TODO(), &overflows); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Error fetching overflows"})
		return
	}
	c.JSON(http.StatusOK, overflows)
}

// Get a specific overflow
func GetOverflow(c *gin.Context) {
	id := c.Param("id")
	var overflow models.Overflow
	err := config.Client.Database("your_db").Collection("overflows").FindOne(context.TODO(), bson.M{"_id": id}).Decode(&overflow)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Overflow not found"})
		return
	}
	c.JSON(http.StatusOK, overflow)
}

// Create a new overflow
func CreateOverflow(c *gin.Context) {
	var overflow models.Overflow
	if err := c.ShouldBindJSON(&overflow); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	_, err := config.Client.Database("your_db").Collection("overflows").InsertOne(context.TODO(), overflow)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Could not create overflow"})
		return
	}
	c.JSON(http.StatusCreated, overflow)
}

// Comment on an overflow
func CommentOnOverflow(c *gin.Context) {
	var comment models.Comment
	if err := c.ShouldBindJSON(&comment); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	id := c.Param("id")
	_, err := config.Client.Database("your_db").Collection("overflows").UpdateOne(
		context.TODO(),
		bson.M{"_id": id},
		bson.M{"$push": bson.M{"comments": comment}},
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Could not comment on overflow"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"message": "Comment added"})
}

// Get user's overflows
func GetUserOverflows(c *gin.Context) {
	username := c.Param("username")
	var overflows []models.Overflow
	cursor, err := config.Client.Database("your_db").Collection("overflows").Find(context.TODO(), bson.M{"username": username})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Error fetching user's overflows"})
		return
	}
	defer cursor.Close(context.TODO())
	if err = cursor.All(context.TODO(), &overflows); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Error fetching user's overflows"})
		return
	}
	c.JSON(http.StatusOK, overflows)
}

File: controllers/overflowController.js
const Overflow = require('../models/Overflow');
const User = require('../models/User');

exports.commentOnOverflow = async (req, res) => {
  const { overflowId } = req.params;
  const { comment } = req.body;

  try {
    const overflow = await Overflow.findById(overflowId);

    if (!overflow) {
      return res.status(404).json({ message: 'overflow not found' });
    }

    const newComment = {
      username: req.user.username,
      text: comment,
      createdAt: new Date(),
    };

    overflow.comments.push(newComment);
    await overflow.save();

    res.status(201).json({ message: 'Comment added', comment: newComment });
  } catch (err) {
    console.error('Server Error:', err);
    res.status(500).json({ message: 'Server error' });
  }
};


exports.getOverflows = async (req, res) => {
  try {
    // const blogs = await Blog.find().populate('author', 'username');
    const overflow = await Overflow.find();
    res.status(200).json(overflow);
  } catch (error) {
    res.status(500).json({ message: 'Server error' });
  }
};

exports.getOverflow = async (req, res) => {
  try {
    const overflow = await Overflow.findById(req.params.id).populate('author', 'username');
    if (!overflow) {
      return res.status(404).json({ message: 'overflow not found' });
    }
    res.status(200).json(overflow);
  } catch (error) {
    res.status(500).json({ message: 'Server error' });
  }
};

exports.createOverflow = async (req, res) => {
  const { title, content } = req.body;

  try {
    const overflow = await Overflow.create({
      title,
      content,
      author: req.user.username,
    });

    res.status(201).json(overflow);
  } catch (error) {
    res.status(500).json({ message: 'Server error' });
  }
};

exports.commentOnOverflow = async (req, res) => {
  const { comment } = req.body;

  try {
    const overflow = await Overflow.findById(req.params.id);

    if (!overflow) {
      return res.status(404).json({ message: 'overflow not found' });
    }

    overflow.comments.push({
      username: req.user.username,
      text: comment.text,
    });

    await overflow.save();

    res.status(201).json(blog);
  } catch (error) {
    res.status(500).json({ message: 'Server error' });
  }
};

exports.getUserOverflows = async (req, res) => {
  try {
    const overflows = await Overflow.find({ author: req.params.username });
    res.json(overflows);
  } catch (error) {
    console.error(error.message);
    res.status(500).send('Server Error');
  }
};
